axoupdater/src/test/helpers.rs

```diff
@@ -0,0 +1,213 @@
+
+/// Generates an install receipt given the specified fields and returns it as a string.
+pub fn install_receipt(
+    app_name: &str,
+    package: &str,
+    owner: &str,
+    binaries: &[String],
+    version: &str,
+    prefix: &str,
+    release_type: &str,
+) -> String {
+    let binaries = binaries
+        .iter()
+        .map(|name| format!(r#""{name}""#))
+        .collect::<Vec<String>>()
+        .join(", ");
+    RECEIPT_TEMPLATE
+        .replace("BINARIES", &binaries)
+        .replace("PACKAGE", package)
+        .replace("OWNER", owner)
+        .replace("APP_NAME", app_name)
+        .replace("INSTALL_PREFIX", &prefix.replace('\\', "\\\\"))
+        .replace("VERSION", version)
+        .replace("RELEASE_TYPE", release_type)
+}
+
+/// Generates an install receipt given the specified fields and writes it to disk at the appropriate location in `config_path`. The path to the new file is returned.
+pub fn write_receipt(
+    app_name: &str,
+    package: &str,
+    owner: &str,
+    binaries: &[String],
+    version: &str,
+    prefix: &Path,
+    config_path: &PathBuf,
+    release_type: &str,
+) -> std::io::Result<PathBuf> {
+    let contents = install_receipt(
+        app_name,
+        package,
+        owner,
+        binaries,
+        version,
+        &prefix.to_string_lossy(),
+        release_type,
+    );
+    let receipt_name = config_path.join(format!("{package}-receipt.json"));
+    std::fs::create_dir_all(config_path)?;
+    std::fs::write(&receipt_name, contents)?;
+
+    Ok(receipt_name)
+}
+
+/// The arguments used for `perform_runtest`.
+pub struct RuntestArgs {
+    /// The name of the app being tested.
+    pub app_name: String,
+    /// The name of the package/workspace being tested. In GitHub terms, this is the "name" of the owner/name repo format.
+    pub package: String,
+    /// The owner of the package being tested. In GitHub terms, this is the "owner" of the owner/name repo format.
+    pub owner: String,
+    /// The path to the executable being tested, usually the one from the `CARGO_BIN_EXE_<name>` environment variable.
+    pub bin: PathBuf,
+    /// A list of all binaries installed by the app being tested.
+    pub binaries: Vec<String>,
+    /// The default path to which the app is installed. For most apps, this is $HOME/.cargo.
+    pub installation_path: PathBuf,
+    /// The arguments taken by the binary being tested. For example, for cargo dist, it's called as `cargo dist selfupdate --skip-init`, so this value is `vec!["dist", "selfupdate", "--skip-init"]`.
+    pub args: Vec<String>,
+    /// The type of release to test; the supported values are `github` (GitHub) and `axodotdev` (Axo Releases).
+    pub release_type: String,
+}
+
+/// Performs an end-to-end run test for `app_name` within `package`.
+/// For detailed information on the arguments, see `RuntestArgs`.
```

Comment:
```suggestion
```suggestion
/// Actually installs your app and runs its updater.
///
/// For detailed information on the arguments, see [`RuntestArgs`][].
```
```

Reply:
(rest of this comment needs some reformatting too)